packages/client/lib/sync/fullsync.ts

```diff
@@ -99,60 +99,119 @@ export class FullSynchronizer extends Synchronizer {
-      count,
-    })
-    this.blockFetcher
-      .on('error', (error: Error) => {
-        this.emit('error', error)
+          if (bulkRequest) {
+            // FIXME
+            // eslint-disable-next-line @typescript-eslint/no-use-before-define
+            blockFetcher.enqueueTask(
+              {
+                first: minCopy,
+                count: numBlocks,
+              },
+              true
+            )
+          } else {
+            data.forEach((value: any) => {
+              const blockNumber: BN = value[1]
+              // FIXME
+              // eslint-disable-next-line @typescript-eslint/no-use-before-define
+              blockFetcher.enqueueTask(
+                {
+                  first: blockNumber,
+                  count: 1,
+                },
+                true
+              )
+            })
+          }
+        }
       })
-      .on('fetched', (blocks: Block[]) => {
-        const first = new BN(blocks[0].header.number)
-        const hash = short(blocks[0].hash())
-        this.config.logger.info(
-          `Imported blocks count=${blocks.length} number=${first.toString(
-            10
-          )} hash=${hash} hardfork=${this.hardfork} peers=${this.pool.size}`
-        )
+      const latest = await this.latest(peer)
+      if (!latest) return false
+      const height = new BN(latest.number)
+      const first = this.chain.blocks.height.addn(1)
+      const count = height.sub(first).addn(1)
+      if (count.lten(0)) return false
+
+      const nextForkBlock = this.config.chainCommon.nextHardforkBlock()
+      if (nextForkBlock) {
+        if (first.gten(nextForkBlock)) {
+          this.config.chainCommon.setHardforkByBlockNumber(first.toNumber())
+          this.hardfork = this.config.chainCommon.hardfork()
+        }
+      }
+
+      this.config.logger.debug(
+        `Syncing with peer: ${peer.toString(true)} height=${height.toString(10)}`
+      )
+
+      this.blockFetcher = new BlockFetcher({
+        config: this.config,
+        pool: this.pool,
+        chain: this.chain,
+        interval: this.interval,
+        first,
+        count,
+        destroyWhenDone: false,
       })
-    await this.blockFetcher.fetch()
-    // TODO: Should this be deleted?
-    // @ts-ignore: error: The operand of a 'delete' operator must be optional
-    delete this.blockFetcher
-    return true
+      const blockFetcher = <BlockFetcher>this.blockFetcher
+      this.blockFetcher
+        .on('error', (error: Error) => {
+          this.emit('error', error)
+        })
+        .on('fetched', (blocks: Block[]) => {
+          const first = new BN(blocks[0].header.number)
+          const hash = short(blocks[0].hash())
+          this.config.logger.info(
+            `Imported blocks count=${blocks.length} number=${first.toString(
+              10
+            )} hash=${hash} hardfork=${this.hardfork} peers=${this.pool.size}`
+          )
+        })
+      await this.blockFetcher.fetch()
+    })
```

Comment:
Haven't looked at this in detail, but just as a general remark, this is a lot of code in bulk, would be great if we can split this up a bit.